working/2365 - primary constructors/feature-specification.md

```diff
@@ -0,0 +1,472 @@
+# Primary Constructors
+
+Author: Erik Ernst
+
+Status: Draft
+
+Version: 1.0
+
+Experiment flag: primary-constructors
+
+This document specifies _primary constructors_. This is a feature that allows
+one constructor and a set of instance variables to be specified in a concise
+form in the header of the declaration. In order to use this feature, the given
+constructor must satisfy certain constraints, e.g., it cannot have a body.
+
+One variant of this feature has been proposed in the [struct proposal][],
+several other proposals have appeared elsewhere, and prior art exists in
+languages like [Kotlin][kotlin primary constructors] and Scala (with
+specification [here][scala primary constructors] and some examples
+[here][scala primary constructor examples]). Many discussions about the 
+feature have taken place in github issues marked with the 
+[primary-constructors label][].
+
+[struct proposal]: https://github.com/dart-lang/language/blob/master/working/extension_structs/overview.md
+[kotlin primary constructors]: https://kotlinlang.org/docs/classes.html#constructors
+[scala primary constructors]: https://www.scala-lang.org/files/archive/spec/2.11/05-classes-and-objects.html#constructor-definitions
+[scala primary constructor examples]: https://www.geeksforgeeks.org/scala-primary-constructor/
+[primary-constructors label]: https://github.com/dart-lang/language/issues?q=is%3Aissue+is%3Aopen+primary+constructor+label%3Aprimary-constructors
+
+## Introduction
+
+Primary constructors is a conciseness feature. It does not provide any new
+semantics at all. It just allows us to express something which is already
+possible in Dart, using a less verbose notation. Here is a simple example:
+
+```dart
+// Current syntax.
+
+class Point {
+  int x;
+  int y;
+  Point(this.x, this.y);
+}
+
+// Same thing using a primary constructor.
+
+class Point(int x, int y);
+```
+
+These examples will serve as an illustration of the proposed syntax, but
+they will also show the semantics of the primary constructor declarations,
+because they will work exactly as the example using the current syntax.
+
+Note that an empty class body, `{}`, can be replaced by `;`.
+
+The basic idea is that a parameter list that occurs just after the class
+name specifies both a constructor declaration and a declaration of one
+instance variable for each formal parameter in said parameter list.
+
+A primary constructor cannot have a body, and it cannot have an initializer
+list (and hence, it cannot have a superinitializer, e.g., `super(...)`, and
```

Comment:
Do you think it would be better to postpone the information about the initializer list as a whole? (The fact that we can't have `super(...)` is just one of the consequences of not having an initializer list.)

The reason why I mentioned this already at this point is that it seems so fundamental that we can't have several syntactic elements (body, initializer list).

But I'll add a few words about the superinitializer when the super parameters are mentioned.